### PR TITLE
Block-diagonal Hessian approximations: KFAC Fisher on multi-class

### PR DESCRIPTION
**Pull request by severinbratus**
KFAC (Kronecker-factored Approximate Curvature): the Fisher information matrix as a block diagonal. This is represented as the `Kron` class, which contains a vector of matrix pairs -- each such pair represents $A_i = E[a_i a_i^T]$ and $G_i = E[g_i g_i^T]$. To obtain the actual approximation of the block approximated by these Kronecker-factors, apply the Kronecker product $A_i \otimes G_i$. Note, here $i$ represents the index of the layer. There are twice as many $A$'s (and equally many $G$'s) as there are layers (since each layer has a weight and a bias). NOTE: Martens & Grosse model the diagonal blocks as accounting for both the weight and the bias at once. This should not be hard to implement, however laplace-torch implements it differently, the same manner it is done in this MR.

The hardest obstacle in obtaining the information needed to get the Kronecker products was that of accessing the intermediate results of forward- and  back-propagation -- the activations $a_i$ and the gradients wrt non-activated linear ouputs of the layer $g_i$. The former is done with `Flux.activate`, the latter through a trick with `pinv`, the pseudo-inverse: since $dW_i = g_i a_{i-1}^T$, we can solve for $g_i$ with $g_i = dW_i (a_{i-1}^{T})^{-1}$.

Note, we are not applying the empirical Fisher information matrix, this is the true Fisher information, since for neural network models with a probabilistic interpretation (such as sigmoid-output binary classifiers or softmax-output multi-class classifiers) we can estimate the expected value of $g$ by computing the gradients for all possible predictions $\hat{y}$, and then weighing these with the corresponding probabilities given by the model. This is expensive, in fact for CIFAR-100 it would be 100 times as expensive (in backward passes) as applying empirical Fisher.

References:
- Martens & Grosse 2015, <https://arxiv.org/abs/1503.05671>
- ASDL, <https://github.com/kazukiosawa/asdl/>
- Kron. product, the vec trick, <https://en.wikipedia.org/wiki/Kronecker_product#Matrix_equations>

TODO:
- [x] Incorporate `Kron` and `get_krons` into the code (possibly refactoring the codebase in the process)
- [x] Alter `get_krons` to work for LL ~~and subnet~~ Laplace
- [x] Write workflow tests: for the `predict` tests to pass, this will probably require procedures:
  - [x] for (pseudo?-)inverting `Kron` (this is easy), and
  - [x] for multiplying `Kron` with a vector via the "vec trick" (should be straight-forward, see refs). Reminder: `Kron` is a Kronecker-factored block-diagonal matrix.
- [ ] ~~[technical, minor] Why are the biases not scaled down in laplace-torch?~~
- [ ] ~~Extend for binary (logit-sigmoid-output) classification~~
- [ ] Write tests comparing to laplace-torch, related to Adelina's testing issue.

NOTE: `nn.jlb` is a network serialized in Julia binary format. `la.H.csv` is the empirical Fisher produced by torch-laplace, `B.csv` is the Block-diagonal (GGN) also by torch-laplace. Note, that these are in row-order, hence they have to be transformed in Julia with some procedures (see `get_mapping_sq`)

### CODE REVIEW

File: src/laplace.jl [hunk 1]
```diff
@@ -1,3 +1,4 @@
+using BlockDiagonals
 using .Curvature
 using Flux
 using Flux.Optimise: Adam, update!
```
Comment (VincentPikand) [inline src/laplace.jl:4]: inconsistent style, are you using `using ...` or `using ...: ...`?

Comment (severinbratus) [reply src/laplace.jl]: We are using both

**Comment by severinbratus** [comment]
Excuse my excessive pipeline crashing -- I did not have the patience for the tests to complete on my machine, + my precommit format hook was misconfigured. Note: subnetwork + KFAC is not sensible, as do get these *slices* of a matrix corresponding to one weight, one  needs two store both Kronecker factors in full, and then one might as well use the all-weights KFAC, since it is more precise. Subnet + KFAC is thus also not implemented in `laplace-torch`: see [this](https://github.com/AlexImmer/Laplace/blob/0faa6fe0cfba0b942a43f90d6cbfc31ea3a8b258/laplace/laplace.py#L23)

**Review by adelinacazacu — COMMENTED** [review]
This is a straight-forward pull request, all of the changes were very well documented on Github. I like the small refactoring that you made, and the implementation is clean and thoroughly tested. Well done! I would also second what @Andrei32Ionescu mentioned, the documentation and comments for the code you added could be a bit more detailed, in the sense of not only specifying what lines x, y and z do, but also offering more technical detail on how this is done.

**Review by Andrei32Ionescu — APPROVED** [review]
Overall, this is great work, well done! Aside from my comment about adding more explanations to the code, I believe we are ready to merge this request.

**Review by MarkArdman — APPROVED** [review]
Not much to complain about here. Very well done with the implementation, nice and clean code, quite like your use of +, * and == operations for Kron. Changes are also well and thoroughly tested. Good job with the implementation and refactoring of the Laplace constructor.

**Comment by adelinacazacu** [comment]
Well written code explanations! I will approve this additionally.

File: src/curvature/functions.jl [hunk 2]
```diff
@@ -236,3 +244,250 @@ function full_batched(curvature::EmpiricalFisher, d::Tuple)
...
+        W_p = reshape(W[:, cur_idx:(cur_idx + sz - 1)], k, p_in, p_out)
+        W_p = permutedims(
+            stack([
+                Q1 * ((transpose(Q1) * W_p[i, :, :] * Q2) .* ldelta_exp) * transpose(Q2) for
+                i in range(1, size(W_p, 1))
+            ]),
+            (3, 1, 2),
+        )
+        W_p = reshape(W_p, k, sz)
+        push!(M, W_p)
```
[the remark this answers is not in the document]
Comment (Andrei32Ionescu) [reply src/curvature/functions.jl]: Thanks for the added explanations!
